manual/manual/tutorials/coreexamples.etex

```diff
@@ -580,6 +580,69 @@ let fixpoint f x =
 the function "f" cannot raise a "Done" exception, which removes an
 entire class of misbehaving functions.
 
+\section{Lazy expressions}
+
+OCaml allows to defer some computation, usually expensive one, until later when 
+we need that expression evaluated. The deferred expression has type "'a Lazy.t", 
+where "'a" is the type of the evaluated expression. 
+
+We use "lazy (expr)" to delay the evaluation of some expression "expr". For 
+example, we can defer the computation of "1+1" until we need the result of that
+expression, i.e. "2". 
+
+\begin{caml_example}{toplevel}
+let lazy_two = lazy (1+1);;
+\end{caml_example}
+
+Note: "'a lazy_t" is a type constructor for keyword "lazy" used by the compiler. 
+One should use type "'a Lazy.t" to work with lazy expressions. 
+
+Also note that the value of "lazy_two" is "<lazy>", which means 
+the expression is not evaluated yet and the result is unknown.
+
+When we finally need the result of the expression, we can call "Lazy.force"  
+on the lazy expression to force the evaluation of the lazy expression. "force" 
+is a function from a standard-library  
+\href{https://caml.inria.fr/pub/docs/manual-ocaml/libref/Lazy.html}
+{module "Lazy"}.
+
+\begin{caml_example}{toplevel}
+  Lazy.force lazy_two;;
+\end{caml_example}
+
+Pay attention that our function call returns plain, unboxed value of the 
+expression. 
+
+We should also note that `Lazy.force` memoizes the result of forced computation. 
+In other words, every subsequent call of "Lazy.force" on the lazy expression is 
+constant time regardless how expensive the initial computation is.
+This can be seen by simply checking the value of the lazy expression that we 
+forced earlier:
+
+\begin{caml_example}{toplevel}
+  lazy_two;;
+\end{caml_example}
+
+We saw that the initial value of "lazy_two" was "<lazy>", but now it is 
+"lazy 2". Thus, forcing "lazy_two" we directly get the result avoiding 
+any computation.
+
+OCaml also provides pattern matching for lazy expressions 
+using "| lazy <pattern>". Lazy expression "v" is matched with 
+"| lazy <pattern>" only if the forced value of "v" matches with "<pattern>". 
+A pattern with keyword "lazy", even if it is a wildcard pattern, always forces 
+the evaluation of the deferred computation.
```

Comment:
I fear that this paragraph plays a bit too fast: it goes from a grammatical description to a technical detail relatively fast:

In general, I am not sure if the grammar description "| lazy <pattern>" fits this chapter (moreover the leading `|` should be removed).

Alternatively, I would start the paragraph by stating that "Lazy patterns provide another way to force a lazy expression"
Then, a simpler illustration for lazy pattern might be:

```OCaml
    let lazy_l = lazy ([1;2] @ [3;4])
    let lazy l = lazy_l;;
```
Chaining that example with nested lazy pattern example would be a good stepping stone.
And finally, at this point,  it would make sense to tell our readers that a lazy pattern always force the evaluation of the deferred computation.